crates/holochain/src/conductor/interface/websocket.rs

```diff
@@ -1,86 +1,119 @@
-use holochain_websocket::ListenerHandle;
-use holochain_websocket::ListenerItem;
+use holochain_websocket::ReceiveMessage;
 use holochain_websocket::WebsocketConfig;
 use holochain_websocket::WebsocketListener;
-use holochain_websocket::WebsocketMessage;
 use holochain_websocket::WebsocketReceiver;
 use holochain_websocket::WebsocketSender;
-use std::convert::TryFrom;
 
 use std::sync::Arc;
 use tokio::sync::broadcast;
 use tokio::task::JoinHandle;
-use tokio_stream::StreamExt;
 use tracing::*;
-use url2::url2;
+
+/// Concurrency count for websocket message processing.
+/// This could represent a significant memory investment for
+/// e.g. app installations, but we also need enough buffer
+/// to accomodate interdependent operations.
+const CONCURRENCY_COUNT: usize = 128;
 
 // TODO: This is arbitrary, choose reasonable size.
+// ERROR TODO XXX (david.b): There is no such thing as backpressure in
+//                           broadcast queues! It'll just start deleting
+//                           items, and giving "Lagged" errors on receivers.
 /// Number of signals in buffer before applying
 /// back pressure.
 pub(crate) const SIGNAL_BUFFER_SIZE: usize = 50;
 /// The maximum number of connections allowed to the admin interface
 pub const MAX_CONNECTIONS: usize = 400;
 
+/// For some reason, the "task manager" doesn't actually shut down tasks...
+/// We have to have this awkward workaround.
+trait TmExt {
+    fn ext_add_task<Fut: std::future::Future<Output = ()> + Send + 'static>(
+        &self,
+        name: &str,
+        f: impl FnOnce() -> Fut + Send + 'static,
+    );
+}
+
+impl TmExt for TaskManagerClient {
+    fn ext_add_task<Fut: std::future::Future<Output = ()> + Send + 'static>(
+        &self,
+        name: &str,
+        f: impl FnOnce() -> Fut + Send + 'static,
+    ) {
+        self.add_conductor_task_ignored(name, move |stop| async move {
+            let task = tokio::task::spawn(f());
+            let _ = stop.await;
+            task.abort();
+            Ok(())
+        });
+    }
+}
+
 /// Create a WebsocketListener to be used in interfaces
-pub async fn spawn_websocket_listener(
-    port: u16,
-) -> InterfaceResult<(
-    ListenerHandle,
-    impl futures::stream::Stream<Item = ListenerItem>,
-)> {
+pub async fn spawn_websocket_listener(port: u16) -> InterfaceResult<WebsocketListener> {
     trace!("Initializing Admin interface");
-    let listener = WebsocketListener::bind_with_handle(
-        url2!("ws://127.0.0.1:{}", port),
+    let listener = WebsocketListener::bind(
         Arc::new(WebsocketConfig::default()),
+        format!("127.0.0.1:{}", port),
     )
     .await?;
-    trace!("LISTENING AT: {}", listener.0.local_addr());
+    trace!("LISTENING AT: {}", listener.local_addr()?);
     Ok(listener)
 }
 
+/// Abort tokio tasks on Drop.
+#[derive(Default)]
+struct TaskList(pub Vec<JoinHandle<()>>);
+impl Drop for TaskList {
+    fn drop(&mut self) {
+        debug!("TaskList Dropped!");
```

Comment:
Is there anything else we can output to give more context?

Reply:
I'm unmotivated to do anything here. Unless there's a bug in the code, you're only going to see this when things are shutting down.